File: lua/bfs.lua
function bfs(start, nodes, fn)
    local frontier = {start}
    local visited = {}
    local next = {}
    local next_visit = function(node)
        local next = {}
        local iter = function (n) return not visited[n] end
        for _, n in ipairs(nodes[node]) do
            if iter(n) then table.insert(next, n) end
        end
        return next
    end

    while 0 < #frontier do
        next = {}
        for _, node in ipairs(frontier) do
            table.insert(visited, node)
            fn(node)
            for _, n in ipairs(next_visit(node)) do
                table.insert(next, n)
            end
        end
        frontier = next
    end
end

return bfs

File: golang/bfs.go
package main

func bfs(start int, nodes map[int][]int, fn func (int)) {
    frontier := []int{start}
    visited := map[int]bool{}
    next := []int{}

    for 0 < len(frontier) {
        next = []int{}
        for _, node := range frontier {
            visited[node] = true
            fn(node)
            for _, n := range bfs_frontier(node, nodes, visited) {
                next = append(next, n)
            }
        }
        frontier = next
    }
}

func bfs_frontier(node int, nodes map[int][]int, visited map[int]bool) []int {
    next := []int{}
    iter := func (n int) bool { _, ok := visited[n]; return !ok }
    for _, n := range nodes[node] {
        if iter(n) {
            next = append(next, n)
        }
    }
    return next
}
